R/run-region-updates.R

```diff
@@ -0,0 +1,90 @@
+  option_list <- list(
+    make_option(c("-v", "--verbose"), action = "store_true", default = FALSE, help = "Print verbose output "),
+    make_option(c("-w", "--werbose"), action = "store_true", default = FALSE, help = "Print v.verbose output "),
+    make_option(c("-q", "--quiet"), action = "store_true", default = FALSE, help = "Print less output "),
+    make_option(c("--log"), type = "character", help = "Specify log file name"),
+    make_option(c("-e", "--exclude"), default = "", type = "character", help = "List of locations to exclude. See include for more details."),
+    make_option(c("-i", "--include"), default = "", type = "character", help = "List of locations to include (excluding all non-specified), comma separated in the format region/subregion or region/*. Case Insensitive. Spaces can be included using quotes - e.g. \"united-states/rhode island, United-States/New York\""),
+    make_option(c("-u", "--unstable"), action = "store_true", default = FALSE, help = "Include unstable locations"),
+    make_option(c("-f", "--force"), action = "store_true", default = FALSE, help = "Run even if data for a region has not been updated since the last run")
+  )
+
+  args <- parse_args(OptionParser(option_list = option_list))
+  return(args)
+}
+
+
+rru_process_locations <- function(regions, args, excludes, includes) {
+  for (location in regions) {
+    if (excludes[region == location$name & subregion == "*", .N] > 0) {
+      futile.logger::flog.debug("skipping location %s as it is in the exclude/* list", location$name)
+      next()
+    }
+    if (includes[, .N] > 0 && includes[region == location$name, .N] == 0) {
+      futile.logger::flog.debug("skipping location %s as it is not in the include list", location$name)
+      next()
+    }
+    if (location$stable || (exists("unstable", args) && args$unstable == TRUE)) {
+      tryCatch(withCallingHandlers({
+                                     update_regional(location,
+                                                     excludes[region == location$name],
+                                                     includes[region == location$name],
+                                                     args$force)
+                                   },
+                                   warning = function(w) {
+                                     futile.logger::flog.warn(w)
+                                   }),
+               error = function(e) {
+                 futile.logger::flog.error(e)
+               }
+      )
+    }else {
+      futile.logger::flog.debug("skipping location %s as unstable", location$name)
+    }
+  }
+}
+
+# only execute if this is the root, passing in regions from region-list.R and the args from the cli interface
+# this bit handles the outer logging wrapping and top level error handling
+if (sys.nframe() == 0) {
+  args <- rru_cli_interface()
+  setup_log_from_args(args)
+  tryCatch(withCallingHandlers(run_regional_updates(regions = regions, args = args),
```

Comment:
@seabbs withCallingHandlers to process the warning into the logger fixes #14  - I'm now getting warnings from stan being spat out
```
WARN [2020-08-27 10:52:31] simpleWarning: Tail Effective Samples Size (ESS) is too low, indicating posterior variances and tail quantiles may be unreliable. 
Running the chains for more iterations may help. See 
http://mc-stan.org/misc/warnings.html#tail-ess

```

Reply:
It does raise that we may want to apply the same inside epinow2 to re-wrap the warnings with something a bit more descriptive
e.g.
run_region_wrapper <- function(region, .......){
withCallingHandlers(EpiNow2::epinow(...), warning=function(w) futile.logger::flog.warn("$s: $s",region, w))
}

Reply:
So whilst not ideal I think this particular warning is okay (stan throws lots of things as warnings which are often indicators of low quality rather than true warnings). We likely need to do some more evaluation of what the best trade-off between compute time and reliable estimates is (as suggested here this will go away if samples is bumped up).

Wrapping the warnings with region info seems like a really good idea.

Reply:
For another PR adding additional logging analysis to highlight where this is happening for and how often etc would be very helpful.

Reply:
I was looking earlier at how you ingest the logs into the cloud log platform - it's something we should do once we move towards a production environment (https://docs.microsoft.com/en-us/azure/azure-monitor/platform/data-sources-custom-logs or similar). If set up right it could draw a graph of frequency over time and tabulate the top offenders.

Reply:
at least it gives us visibility of the warnings now even if we choose to ignore them!

Reply:
That does sound like a good idea.